### ISSUE TITLE
AppInsights Light-up missing JavaScriptSnippet?

### ISSUE THREAD

**Issue by onovotny** [Microsoft/ApplicationInsights-aspnetcore#568]
Using an ASP.NET Core 2 app running on App Services (with App Insights configured), I expected the “light up” to just-work.

The problem though is that if I want to use the client-side javascript snippets, I get errors due to:
`No service for type 'Microsoft.ApplicationInsights.AspNetCore.JavaScriptSnippet' has been registered.`

I also note that there’s conflicting information as to whether adding:

`@inject Microsoft.ApplicationInsights.AspNetCore.JavaScriptSnippet JavaScriptSnippet`
And `@Html.Raw(JavaScriptSnippet.FullScript)`

Are necessary or should be removed. I didn’t seem to see the AI script unless I added them. 

If this is supposed to be all automatic, shouldn’t that be included?

What finally resolved the issue was adding a `UseApplicationInsights()` call in my `Program.cs`. I’m using the regular template / `WebHost.CreateDefaultBuilder` there. Adding that call seems like going against the “light up,” it-just-works approach.

So which is it? Is that needed? Should the javascript stuff just work?

**Comment by glennc** [comment]
Need some extra information here.

1. When using lightup on App Service the javascript is inserted for you when it is working. We inject a tag helper that adds the javascript (https://github.com/aspnet/AzureIntegration/blob/f4236a1dde10c0d2ef790f75a5a8e4a25d582f35/src/Microsoft.AspNetCore.ApplicationInsights.HostingStartup/JavaScriptSnippetTagHelperComponent.cs)
2. Did you want to do this because it wasn't there? In which case lightup wasn't working and we can diagnose that, or did you want to do something to configure/change it?

I need to make sure of the versions of everything you are using as well. metapackage version, .NET Core version, etc :).

**Comment by onovotny** [comment]
I didn't see the any script on the page itself. the `APPINSIGHTS_INSTRUMENTATIONKEY` was set though.

I added the `UseAppliacationInsights()` because the injected component was missing unless I did that.

Here's my current set of references. Should be the latest:
https://github.com/onovotny/SignService/blob/master/src/SignService/SignService.csproj#L28-L38

**Comment by glennc** [comment]
Do you have the app insights site extension installed?

**Comment by onovotny** [comment]
I don't think I originally did, though I do now. The docs weren't clear on whether that was required or not or if just setting `APPINSIGHTS_INSTRUMENTATIONKEY` was enough for some data while the site extension provides additional data.

**Comment by glennc** [comment]
The site extension is required to get Javascript and such. Can you remove the code from your app that uses application insights types and re-deploy with the site extension installed?

**Comment by glennc** [comment]
Also, link me to the docs you are referencing so I can clarify their wording. Specifically the site extension sets two more env vars that add the ASP.NET Core logger, the javascript tag helper, etc.

**Comment by onovotny** [comment]
Here's one:
https://github.com/Microsoft/ApplicationInsights-aspnetcore/wiki/Getting-Started-for-a-ASP.NET-CORE2.0-WebApp

In the docs, it says create the resource, add the instrumentationkey. It also mentions "adding JS instrumentation to the viewimports and layout.

This doc set also seems to indicate something similar:
https://docs.microsoft.com/en-us/azure/application-insights/app-insights-dotnetcore-quick-start
Create the resource, add the key... Nothing in there mentions the site extension either.

Nothing in there mentions the site extension. Also, nothing in there mentions the extra code needed for the snapshot collector (will that be rolled in and automatic eventually)?



Do I need to keep or get rid of the `@Html.Raw` on the `_Layout.cshtml` page as well?

**Comment by glennc** [comment]
Yes, remove the code that adds the snippet now. You shouldn't need any references to app insights in your app.

Yes, the snapshot collector stuff will be rolled into the extension and you will get it by updating the extension to a new version. I am not sure what the status of that is right now, but I don't think it would be far away. @cweining would know more about that.

**Comment by onovotny** [comment]
Removing the javascript snippet, the html.raw and the useapplicationinsights now has no script getting generated.

https://oren-signservice-dev.azurewebsites.net/

View source on that and I don't see it :)

**Comment by glennc** [comment]
Does your app have any env vars talking about HostingStartup in it? If so, what are they.

If the site extension is actually working you should see some Application Insights specific values is ASPNETCORE_HOSTINGSTARTUPASSEMBLIES. You should also have a DOTNET_ADDITIONAL_DEPS env var I think, since we always set it.

**Comment by glennc** [comment]
If these are not set to the right things then the site extension isn't working for some reason. Failed to install properly perhaps. I would first go for a re-install of that. Then we can look at going through the logs from the extension.

If those vars are set then the problem is in the HostingStartup code somewhere and we will have to debug it differently.

**Comment by onovotny** [comment]
Looking at the Kudu env:
ASPNETCORE_HOSTINGSTARTUPASSEMBLIES = Microsoft.AspNetCore.AzureAppServices.HostingStartup
DOTNET_ADDITIONAL_DEPS = D:\Program Files (x86)\dotnet\additionalDeps\Microsoft.AspNetCore.AzureAppServices.HostingStartup\

**Comment by glennc** [comment]
OK, the App Insights site extension isn't working. Can you remove it and then install it again?

**Comment by glennc** [comment]
Those env vars are just to configure the App Service specific log provider that logs to the App Service log location an enables streaming logs.

**Comment by cijothomas** [comment]
@onovotny Please try to remove the ai extension and re-install. If lighting up Ai from AI Extension for webapps, then no need of making any code changes/cshtml changes for js.
(I believe we don't have proper documentation for Web App light up - will add one soon)

**Comment by xt0rted** [comment]
I updated my site from 2.0 to 2.1, deployed to a new instance to test, and noticed the AI JS isn't being emitted on my page, nor is there any server telemetry being tracked.

Upon further investigation I found that my production site isn't working any more either. On February 24th I removed the explicit `@Html.Raw(JavaScriptSnippet.FullScript)` when I realized that the JS was included twice. At some point between then and now the JS stopped being written out but I'm not sure why.

To better test this I created a new 2.1 project:
1. New Web Application (Model-View-Controller)
2. Installed the AI hosting startup package
3. Published to a new app service instance w/ app insights

Testing at this point showed there was no JS or server telemetry being tracked. I then installed the AI site extension and restarted the site. I'm still not getting the JS or server telemetry. These are the env var values that I'm seeing.

Name | Value
-- | --
ASPNETCORE_HOSTINGSTARTUPASSEMBLIES | StartupBootstrapper;Microsoft.AspNetCore.AzureAppServices.HostingStartup;Microsoft.AspNetCore.Server.IISIntegration
DOTNET_ADDITIONAL_DEPS | D:\Program Files (x86)\dotnet\additionalDeps\Microsoft.AspNetCore.ApplicationInsights.HostingStartup\;D:\home\SiteExtensions\Microsoft.ApplicationInsights.AzureWebSites\2.6.1\core\additionalDeps;D:\home\SiteExtensions\Microsoft.AspNetCore.AzureAppServices.SiteExtension\additionalDeps\Microsoft.AspNetCore.AzureAppServices.HostingStartup\;D:\Program Files (x86)\dotnet\additionalDeps\Microsoft.AspNetCore.AzureAppServices.HostingStartup\

The only way I can get the AI hosting startup package to work is by adding an explicit env var for `ASPNETCORE_HOSTINGSTARTUPASSEMBLIES` with the value `Microsoft.AspNetCore.AzureAppServices.HostingStartup` to my web.config. Once that's in place the JS starts showing up on the site and live metrics start working (if you update AI to 2.2.0+ that is).
